functions/src/place-bet.ts

```diff
@@ -1,146 +1,140 @@
-
-  if (value !== undefined && !isFinite(value))
-    throw new APIError(400, 'Invalid value')
-
-  // run as transaction to prevent race conditions
-  return await firestore
-    .runTransaction(async (transaction) => {
-      const userDoc = firestore.doc(`users/${bettor.id}`)
-      const userSnap = await transaction.get(userDoc)
-      if (!userSnap.exists) throw new APIError(400, 'User not found')
-      const user = userSnap.data() as User
-
-      const contractDoc = firestore.doc(`contracts/${contractId}`)
-      const contractSnap = await transaction.get(contractDoc)
-      if (!contractSnap.exists) throw new APIError(400, 'Invalid contract')
-      const contract = contractSnap.data() as Contract
+const bodySchema = z.object({
+  contractId: z.string(),
+  amount: z.number().gte(1),
+})
 
-      const { closeTime, outcomeType, mechanism, collectedFees, volume } =
-        contract
-      if (closeTime && Date.now() > closeTime)
-        throw new APIError(400, 'Trading is closed')
+const binarySchema = z.object({
+  outcome: z.enum(['YES', 'NO']),
+})
 
-      const yourBetsSnap = await transaction.get(
-        contractDoc.collection('bets').where('userId', '==', bettor.id)
-      )
-      const yourBets = yourBetsSnap.docs.map((doc) => doc.data() as Bet)
+const freeResponseSchema = z.object({
+  outcome: z.string(),
+})
 
-      const loanAmount = 0 // getLoanAmount(yourBets, amount)
-      if (user.balance < amount) throw new APIError(400, 'Insufficient balance')
+const numericSchema = z.object({
+  outcome: z.string(),
+  value: z.number(),
+})
 
-      if (outcomeType === 'FREE_RESPONSE') {
-        const answerSnap = await transaction.get(
-          contractDoc.collection('answers').doc(outcome)
+export const placeBet = newEndpoint(['POST'], async (req, [bettor, _]) => {
+  const { amount, contractId } = validate(bodySchema, req.body)
+
+  const result = await firestore.runTransaction(async (trans) => {
+    const userDoc = firestore.doc(`users/${bettor.id}`)
+    const userSnap = await trans.get(userDoc)
+    if (!userSnap.exists) throw new APIError(400, 'User not found.')
+    const user = userSnap.data() as User
+    if (user.balance < amount) throw new APIError(400, 'Insufficient balance.')
+
+    const contractDoc = firestore.doc(`contracts/${contractId}`)
+    const contractSnap = await trans.get(contractDoc)
+    if (!contractSnap.exists) throw new APIError(400, 'Contract not found.')
+    const contract = contractSnap.data() as Contract
+
+    const loanAmount = 0
+    const { closeTime, outcomeType, mechanism, collectedFees, volume } =
+      contract
+    if (closeTime && Date.now() > closeTime)
+      throw new APIError(400, 'Trading is closed.')
+
+    const newBetDoc = contractDoc.collection('bets').doc()
+    const {
+      newBet,
+      newPool,
+      newTotalShares,
+      newTotalBets,
+      newTotalLiquidity,
+      newP,
+    } = await (async (): Promise<BetInfo> => {
```

Comment:
Might want to extract this logic into a new (named) function

Reply:
Unfortunately, it's kind of awkward to do that, since each branch relies on, like, validating additional parameters on the request body, and one branch looks up stuff using the Firestore transaction, so it feels really poorly factored to break it out. The broken out function would take like eight random parameters.